File: Tests/ConnectorControlCoreTests/ToolNoteTests.swift
import XCTest
@testable import ConnectorControlCore

final class ToolNoteTests: XCTestCase {
    func testMissingToolNoteCarriesTheInstallLinkAndCommand() {
        let note = ToolNote.make(tool: .npx, status: .notFound)
        XCTAssertEqual(note?.text, "npx wasn’t found, so Claude Desktop won’t be able to start this connector.")
        XCTAssertEqual(note?.linkTitle, "Install Node.js")
        XCTAssertEqual(note?.linkURL.absoluteString, "https://nodejs.org/en/download")
        XCTAssertEqual(note?.installCommand, "brew install node")
        let uv = ToolNote.make(tool: .uv, status: .notFound)
        XCTAssertEqual(uv?.text, "uv wasn’t found, so Claude Desktop won’t be able to start this connector.")
        XCTAssertEqual(uv?.linkTitle, "Install uv")
        XCTAssertEqual(uv?.linkURL.absoluteString, "https://docs.astral.sh/uv/getting-started/installation/")
        XCTAssertEqual(uv?.installCommand, "brew install uv")
        XCTAssertNil(note?.advice, "the advice line belongs to the shell-only state")
        XCTAssertNil(ToolNote.make(tool: .node, status: .found(path: "/usr/local/bin/node", version: "22.11.0")))
        XCTAssertNil(ToolNote.make(tool: .node, status: nil), "unknown is not a problem yet")
    }

    func testShellOnlyNoteNamesThePathAndAdvisesAVisibleInstall() {
        let path = "/Users/me/.nvm/versions/node/v22.11.0/bin/npx"
        let note = ToolNote.make(tool: .npx, status: .foundInShellOnly(path: path, version: "10.9.2"))
        XCTAssertEqual(note?.text, "npx is at \(path) in your shell, but Claude Desktop launches connectors with its own PATH and may not see it there.")
        XCTAssertEqual(note?.advice, "Install it with Homebrew or the official installer, which put it where Claude Desktop looks; version managers such as nvm install elsewhere.")
        XCTAssertEqual(note?.linkTitle, "Install Node.js")
        XCTAssertEqual(note?.installCommand, "brew install node",
                       "Homebrew installs into the PATH floor, so this fix clears the note")
        let uvx = ToolNote.make(tool: .uvx, status: .foundInShellOnly(path: "/Users/me/.local/bin/uvx", version: nil))
        XCTAssertEqual(uvx?.installCommand, "brew install uv")
        XCTAssertEqual(uvx?.advice, ToolNote.shellOnlyAdvice)
    }

    func testStatusTextForEachState() {
        XCTAssertEqual(ToolNote.statusText(nil), "Checking…")
        XCTAssertEqual(ToolNote.statusText(.found(path: "/x/npx", version: "10.9.2")), "10.9.2")
        XCTAssertEqual(ToolNote.statusText(.found(path: "/x/npx", version: nil)), "Found")
        XCTAssertEqual(ToolNote.statusText(.foundInShellOnly(path: "/x/npx", version: "10.9.2")), "Not visible to Claude Desktop")
        XCTAssertEqual(ToolNote.statusText(.notFound), "Not found")
    }

    /// Ordering, name mapping, and lookup — behavior StringCatalogTests
    /// doesn't cover (it pins strings, not enum order or lookup logic).
    func testOrderNameMappingAndFamilyLookup() {
        XCTAssertEqual(Tool.allCases, [.npx, .node, .uvx, .uv])
        XCTAssertEqual(Tool.allCases.map(\.name), ["npx", "node", "uvx", "uv"])
        XCTAssertEqual(Tool(name: "NPX"), .npx)
        XCTAssertNil(Tool(name: "python"))
        XCTAssertEqual(Tool.npx.family, .nodeJS)
        XCTAssertEqual(Tool.node.family, .nodeJS)
        XCTAssertEqual(Tool.uvx.family, .uv)
        XCTAssertEqual(Tool.uv.family, .uv)
    }

    func testRowWarningIsTheRowTooltipForEachStatus() {
        XCTAssertNil(ToolNote.rowWarning(tool: .npx, status: nil),
                     "unknown must never flash a glyph before the probe has answered")
        XCTAssertNil(ToolNote.rowWarning(tool: .npx, status: .found(path: "/opt/homebrew/bin/npx", version: "10.9.2")))
        XCTAssertNil(ToolNote.rowWarning(tool: .npx, status: .found(path: "/opt/homebrew/bin/npx", version: nil)))
        XCTAssertEqual(ToolNote.rowWarning(tool: .npx, status: .notFound),
                       "Needs npx, which wasn’t found. Edit to see how to install it.")
        XCTAssertEqual(ToolNote.rowWarning(tool: .uvx, status: .notFound),
                       "Needs uvx, which wasn’t found. Edit to see how to install it.")
        XCTAssertEqual(ToolNote.rowWarning(tool: .node,
                                           status: .foundInShellOnly(path: "/Users/me/.nvm/bin/node", version: "22.11.0")),
                       "Needs node, which Claude Desktop may not see. Edit to see how to fix it.")
        XCTAssertEqual(ToolNote.rowMissingText(.uv),
                       "Needs uv, which wasn’t found. Edit to see how to install it.")
        XCTAssertEqual(ToolNote.rowShellOnlyText(.uv),
                       "Needs uv, which Claude Desktop may not see. Edit to see how to fix it.")
    }
}

File: windows/tests/ConnectorControl.Core.Tests/ToolNoteTests.cs
namespace ConnectorControl.Core.Tests;

public class ToolNoteTests
{
    [Fact]
    public void MissingToolNoteCarriesTheInstallLinkAndCommand()
    {
        var note = ToolNote.Make(Tool.Npx, ToolStatus.NotFound);
        Assert.NotNull(note);
        Assert.Equal("npx wasn’t found, so Claude Desktop won’t be able to start this connector.", note.Text);
        Assert.Equal("Install Node.js", note.LinkTitle);
        Assert.Equal("https://nodejs.org/en/download", note.LinkUrl);
        Assert.Equal("winget install OpenJS.NodeJS.LTS", note.InstallCommand);
        var uv = ToolNote.Make(Tool.Uv, ToolStatus.NotFound);
        Assert.NotNull(uv);
        Assert.Equal("uv wasn’t found, so Claude Desktop won’t be able to start this connector.", uv.Text);
        Assert.Equal("Install uv", uv.LinkTitle);
        Assert.Equal("https://docs.astral.sh/uv/getting-started/installation/", uv.LinkUrl);
        Assert.Equal("winget install astral-sh.uv", uv.InstallCommand);
        Assert.Null(ToolNote.Make(Tool.Node, new ToolStatus(@"C:\Program Files\nodejs\node.exe", "22.11.0")));
        Assert.Null(ToolNote.Make(Tool.Node, null));   // unknown is not a problem yet
    }

    [Fact]
    public void StatusTextForEachState()
    {
        Assert.Equal("Checking…", ToolNote.StatusText(null));
        Assert.Equal("10.9.2", ToolNote.StatusText(new ToolStatus(@"C:\x\npx.cmd", "10.9.2")));
        Assert.Equal("Found", ToolNote.StatusText(new ToolStatus(@"C:\x\npx.cmd", null)));
        Assert.Equal("Not found", ToolNote.StatusText(ToolStatus.NotFound));
    }

    [Fact]
    public void StringsAndOrderMatchTheSpec()
    {
        Assert.Equal([Tool.Npx, Tool.Node, Tool.Uvx, Tool.Uv], ToolInfo.All.ToArray());
        Assert.Equal(["npx", "node", "uvx", "uv"], ToolInfo.All.Select(ToolInfo.Name).ToArray());
        Assert.Equal(Tool.Npx, ToolInfo.Parse("NPX"));
        Assert.Null(ToolInfo.Parse("python"));
        Assert.Equal(ToolFamily.NodeJs, ToolInfo.Family(Tool.Npx));
        Assert.Equal(ToolFamily.NodeJs, ToolInfo.Family(Tool.Node));
        Assert.Equal(ToolFamily.Uv, ToolInfo.Family(Tool.Uvx));
        Assert.Equal(ToolFamily.Uv, ToolInfo.Family(Tool.Uv));
        Assert.Equal("or run", ToolNote.OrRun);
        Assert.Equal("Tools", ToolNote.SettingsHeader);
        Assert.Equal("Connectors that run through npx, node, uvx or uv need them installed where Claude Desktop can find them.", ToolNote.SettingsCaption);
        Assert.Equal(".COM;.EXE;.BAT;.CMD", ToolProbe.DefaultPathExt);
    }

    [Fact]
    public void RowWarningIsTheRowTooltipForEachStatus()
    {
        Assert.Null(ToolNote.RowWarning(Tool.Npx, null));   // unknown must never flash a glyph before the probe has answered
        Assert.Null(ToolNote.RowWarning(Tool.Npx, new ToolStatus(@"C:\Program Files\nodejs\npx.cmd", "10.9.2")));
        Assert.Null(ToolNote.RowWarning(Tool.Npx, new ToolStatus(@"C:\Program Files\nodejs\npx.cmd", null)));
        Assert.Equal("Needs npx, which wasn’t found. Edit to see how to install it.", ToolNote.RowWarning(Tool.Npx, ToolStatus.NotFound));
        Assert.Equal("Needs uvx, which wasn’t found. Edit to see how to install it.", ToolNote.RowWarning(Tool.Uvx, ToolStatus.NotFound));
        Assert.Equal("Needs uv, which wasn’t found. Edit to see how to install it.", ToolNote.RowMissingText(Tool.Uv));
        Assert.Equal("Needs node, which wasn’t found. Edit to see how to install it.", ToolNote.RowMissingText(Tool.Node));
    }
}
